actors/builtin/miner/deadline_state_test.go

```diff
@@ -0,0 +1,379 @@
+				dl,
+				sectors,
+				sectorSize,
+				partitionSize,
+				bf(9),       // faults
+				bf(),        // recovering
+				bf(1, 3, 6), // terminations
+				bf(),        // posts
+
+				bf(1, 2, 3, 4),
+				bf(5, 6, 7, 8),
+				bf(9),
+			)
+		}
+
+		// Try to compact a partition with early terminations.
+		{
+			_, _, _, err := dl.RemovePartitions(store, bf(0), quantSpec)
+			require.Error(t, err, "should have failed to compact a partition with early terminations")
+		}
+
+		// Pop early terminations
+		{
+			earlyTerminations, more, err := dl.PopEarlyTerminations(store, 100, 100)
+			require.NoError(t, err)
+			assert.False(t, more)
+			assert.Equal(t, uint64(2), earlyTerminations.PartitionsProcessed)
+			assert.Equal(t, uint64(3), earlyTerminations.SectorsProcessed)
+			assert.Len(t, earlyTerminations.Sectors, 1)
+			assertBitfieldEquals(t, earlyTerminations.Sectors[15], 1, 3, 6)
+
+			// Popping early terminations doesn't affect the terminations bitfield.
+			assertDeadlineState(t, rt,
+				dl,
+				sectors,
+				sectorSize,
+				partitionSize,
+				bf(9),       // faults
+				bf(),        // recovering
+				bf(1, 3, 6), // terminations
+				bf(),        // posts
+
+				bf(1, 2, 3, 4),
+				bf(5, 6, 7, 8),
+				bf(9),
+			)
+		}
+
+		// Try to compact no partitions
+		{
+			live, dead, removedPower, err := dl.RemovePartitions(store, bf(), quantSpec)
+			require.NoError(t, err, "should not have failed to compact no partitions")
+			require.True(t, removedPower.IsZero())
+			assertBitfieldEquals(t, live)
+			assertBitfieldEquals(t, dead)
+
+			// Popping early terminations doesn't affect the terminations bitfield.
+			assertDeadlineState(t, rt,
+				dl,
+				sectors,
+				sectorSize,
+				partitionSize,
+				bf(9),       // faults
+				bf(),        // recovering
+				bf(1, 3, 6), // terminations
+				bf(),        // posts
+
+				bf(1, 2, 3, 4),
+				bf(5, 6, 7, 8),
+				bf(9),
+			)
+		}
+
+		// Remove some partitions
+		{
+			live, dead, removedPower, err := dl.RemovePartitions(store, bf(0), quantSpec)
+			require.NoError(t, err, "should have removed partitions")
+			assertBitfieldEquals(t, live, 2, 4)
+			assertBitfieldEquals(t, dead, 1, 3)
+			livePower := miner.PowerForSectors(sectorSize, selectSectors(t, sectors, live))
+			require.True(t, livePower.Equals(removedPower))
+
+			assertDeadlineState(t, rt,
+				dl,
+				sectors,
+				sectorSize,
+				partitionSize,
+				bf(9), // faults
+				bf(),  // recovering
+				bf(6), // terminations
+				bf(),  // posts
+
+				bf(5, 6, 7, 8),
+				bf(9),
+			)
+		}
+
+		// Try to compact a partition with faulty sectors.
+		{
+			_, _, _, err := dl.RemovePartitions(store, bf(1), quantSpec)
+			require.Error(t, err, "should have failed to compact a partition with faults")
+		}
+
+		// Try to compact a missing partition.
+		{
+			_, _, _, err := dl.RemovePartitions(store, bf(2), quantSpec)
+			require.Error(t, err, "should have failed to compact missing partition")
+		}
+	})
```

Comment:
This is borderline too much. We should do some work to make writing these tests more concise, because we'll need more of them. Refactoring the deadline to have a more coherent API and less direct state access may help.

Reply:
I've introduced a state checking abstraction and deduped the tests. How does it look now?

Reply:
I'd still like to push recording faults down into a deadline function, but that's going to need a new PR.